### PR TITLE
Improved DALI support for training and inference

### PR DESCRIPTION
**Pull request by pkashinkunti**
Improve DALI dataloaders for training and inference. Full pipeline done through DALI as opposed to DALI + PyTorch. 

PyTorch dataloader is still the default unless `--with-dali` is added. This is because the current DALI COCOReader class hardcodes COCO categories and will not work with other COCO style datasets.

### CODE REVIEW

File: retinanet/dali.py [hunk 1]
```diff
@@ -83,50 +100,71 @@ def __repr__(self):
         ])
 
     def __len__(self):
-        return ceil(len(self.local_ids) / self.batch_size)
+        return ceil(len(self.ids) // self.world / self.batch_size)
```
Comment (florischabert) [inline retinanet/dali.py]: Does that work when len(ids) % self.word != 0?

File: retinanet/dali.py [hunk 2]
```diff
@@ -83,50 +100,71 @@ def __repr__(self):
...
+            dali_data, dali_boxes, dali_labels, dali_ids, dali_attrs, dali_resize_img = self.pipe.run()
+
+            for l in range(len(dali_boxes)):
+                num_detections.append(dali_boxes.at(l).shape[0])
+
+            pyt_targets = -1 * torch.ones([len(dali_boxes), max(max(num_detections),1), 5])
+
             for batch in range(self.batch_size):
                 id = int(dali_ids.at(batch)[0])
-                if id < 0: break
```
Comment (florischabert) [inline retinanet/dali.py]: What happens for the last batch if we have less items left that batch_size?

File: retinanet/dali.py [hunk 3]
```diff
@@ -83,50 +100,71 @@ def __repr__(self):
...
-            data, ratios, ids = [], [], []
-            dali_data, dali_ids = self.pipe.run()
+
+            data, ratios, ids, num_detections = [], [], [], []
+            dali_data, dali_boxes, dali_labels, dali_ids, dali_attrs, dali_resize_img = self.pipe.run()
+
+            for l in range(len(dali_boxes)):
+                num_detections.append(dali_boxes.at(l).shape[0])
+
+            pyt_targets = -1 * torch.ones([len(dali_boxes), max(max(num_detections),1), 5])
```
Comment (florischabert) [inline retinanet/dali.py]: If there was no detection for any target, is having a -1 filled detection potentially an issue?

Comment (pkashinkunti) [reply retinanet/dali.py]: Won't it be ignored per https://github.com/NVIDIA/retinanet-examples/blob/master/retinanet/model.py#L138 since the label value is -1?

This is also similar as far as I see with what we do for the regular dataloader: https://github.com/NVIDIA/retinanet-examples/blob/master/retinanet/data.py#L101
https://github.com/NVIDIA/retinanet-examples/blob/master/retinanet/data.py#L111

**Comment by florischabert** [comment]
Let's resolve the last batch issues in a subsequent PR
